gym/spaces/graph.py

```diff
@@ -0,0 +1,260 @@
+        dimension (denoted n) is the number of edges.
+        If edges is None, then edge_links must also be None.
+
+        ``edge_links`` must be a nx2 sized array of ints, where edge_links.max()
+        is not to be equal or larger than the size of the first dimension of nodes, and
+        edge_links.min() is not to be smaller than 0.
+        """
+        self.nodes = nodes
+        self.edges = edges
+        self.edge_links = edge_links
+
+    def __repr__(self) -> str:
+        """A string representation of this graph.
+
+        The representation will include nodes, edges, and edge_links
+
+        Returns:
+            The information in this graph.
+        """
+        return f"GraphObj(nodes: \n{self.nodes}, \n\nedges: \n{self.edges}, \n\nedge_links\n{self.edge_links})"
+
+    def __eq__(self, other) -> bool:
+        """Check whether `other` is equivalent to this instance."""
+        if not isinstance(other, GraphObj):
+            return False
+        if np.any(self.nodes != other.nodes):
+            return False
+        if self.edges is not None:
+            if other.edges is None:
+                print("fail1")
+                return False
+            if np.all(self.edges != other.edges):
+                print("fail2")
+                return False
+            if other.edge_links is None:
+                print("fail3")
+                return False
+            if np.all(self.edge_links != other.edge_links):
+                print("fail4")
+                return False
+        else:
+            if other.edges is not None:
+                print("fail6")
+                return False
+            if other.edge_links is not None:
+                print("fail7")
+                return False
+
+        return True
+
+
+class Graph(Space):
+    r"""A dictionary representing graph spaces with `node_features`, `edge_features` and `edge_links`.
+
+    Example usage::
+
+        self.observation_space = spaces.Graph(node_space=space.Box(low=-100, high=100, shape=(3,)), edge_space=spaces.Discrete(3))
+    """
+
+    def __init__(
+        self,
+        node_space: Union[Box, Discrete],
+        edge_space: Union[None, Box, Discrete],
+        seed: Optional[Union[int, seeding.RandomNumberGenerator]] = None,
+    ):
+        r"""Constructor of :class:`Graph`.
+
+        The argument ``node_space`` specifies the base space that each node feature will use.
+        This argument must be either a Box or Discrete instance.
+
+        The argument ``edge_space`` specifies the base space that each edge feature will use.
+        This argument must be either a Box or Discrete instance.
+
+        Args:
+            node_space (Union[Box, Discrete]): space of the node features.
+            edge_space (Union[None, Box, Discrete]): space of the node features.
+            seed: Optionally, you can use this argument to seed the RNG that is used to sample from the space.
+        """
+        self._np_random = None
```

Comment:
Why is this necessary, should this not be passed to the `super().__init__` which will handle this?

Reply:
I'm not completely sure about, I kinda just used the Box space as a template for building thid space, but I'll have a look again

Reply:
It appears I stole template code from an older version of Box, I'll amend things.